README.md
This is the source directory of **DEFINER** ([gitslice/definer](https://bitbucket.org/liyistc/gitslice/commits/tag/definer)) and **CSLICER** ([gitslice/cslicer](https://bitbucket.org/liyistc/gitslice/commits/tag/cslicer)), prototype implementations of a collection of automated semantic history slicing algorithms. The tools currently work for Java projects hosted in Git repositories.

---

## License

The license is in LICENSE.txt

## Introduction

DEFINER is a semantic slicing tool for Git reposiotries under active development by the Software Engineering group at the University of Toronto.
The goal of semantic slicing is to identify a subset of change sets (commits) that implementing a software functionality (e.g., feature, enhancement, or bug fix). The applications include code migration, branch refactoring, etc.
DEFINER currently works only for Java repositories.

Another implementation which is based on an alternative semantic slicing algorithm -- CSLICER is now residing in the master branch.

## Dependencies

* Maven  
[Apache Maven](http://maven.apache.org) is used for dependency control. To build the project, first download and install Maven following the instructions [here](http://maven.apache.org/download.cgi) (version >= 3.2). Make sure environment variables ```M2_HOME```, ```M2```, and ```JAVA_HOME``` are set correctly according to the instructions.
All required dependencies will be fetched and installed by Maven automatically.

* ChangeDistiller  
We use [ChangeDistiller](https://bitbucket.org/sealuzh/tools-changedistiller/wiki/Home) for categorizing and analyzing changes.
A local copy of ChangeDistiller-SNAPSHOT-0.0.1 is included in the lib directory.

* Daikon (optional)   
We use [Daikon](https://plse.cs.washington.edu/daikon) for dynamic invariant generation. A local copy of Daikon is included in the lib directory.

## Install DEFINER

Requirements:

* JDK 1.8
* Maven 3.0 or later
* Internet connection for first build (to fetch all Maven and DEFINER dependencies)

Get the source and external dependencies
```shell
git clone https://github.com/ntu-SRSLab/CSlicer.git cslicer
cd cslicer
git submodule init
git submodule update
```

Maven build goals:

* Clean: ```mvn clean```
* Compile: ```mvn compile```
* Run tests: ```mvn test```
* Create JAR: ```mvn package```
* Install JAR in M2 cache: ```mvn install```
* Build javadocs: ```mvn javadoc:javadoc```

Build options:

* Use ```-DskipTests``` to skip tests

## Run DEFINER ##

Find executable jar files at ```definer/target``` and all dependencies at ```definer/target/lib```. Run DEFINER using
 
```shell
java -jar cslicer-XXX-jar-with-dependencies.jar -c <CONFIG_FILE_PATH>
```

The target project configuration file specifies the location, target commits (history), path to source code root, etc. An example of a configuration file (with extension .properties) is as follows.

```properties
# This is an example project configuration file project.properties.
# Assume example repository is located within <EXAMPLE_DIRECTORY>
# Path to target repository (required)
repoPath = <EXAMPLE_DIRECTOR>/commons-collections-examples/IterableUtilsTest#indexOf/commons-collections/.git
# Target history length (optional)
historyLength = 97
# History ending commit (required)
endCommit = dcf3df80
# Path to the pom.xml build script (required)
buildScriptPath = <EXAMPLE_DIRECTOR>/commons-collections-examples/IterableUtilsTest#indexOf/commons-collections/pom.xml
# Test scope settings: 'indexOf' method in 'IterableUtilsTest' class (required)
testScope = IterableUtilsTest#indexOf
# Path to the configuration file used by Daikon (optional)
daikonConfig = /home/polaris/Desktop/123/example-settings.txt
# Path to Daikon jar file (required)
daikonJar = /home/polaris/Desktop/daikonparent/daikon-5.2.22/daikon.jar
# Path to Chicory jar (required by Daikon)
chicoryPath = /home/polaris/Desktop/daikonparent/daikon-5.2.22/java/ChicoryPremain.jar
# Excludes changes in the following files (optional)
excludePaths = pom.xml,src/changes/changes.xml,.gitignore,CONTRIBUTING.md,NOTICE.txt,README.md,README.txt,TODO.txt
```

See detailed usage by entering the command

```shell
java -jar cslicer-XXX-jar-with-dependencies.jar -h
```

## Publications ##
[Semantic Slicing of Software Version Histories](https://www.ntu.edu.sg/home/yi_li/posts/paper-accepted-by-tse/)   
Yi Li, Chenguang Zhu, Julia Rubin and Marsha Chechik   
In Transactions on Software Engineering (TSE), 44(2):182–201

[Precise Semantic History Slicing through Dynamic Delta Refinement](https://www.ntu.edu.sg/home/yi_li/posts/paper-accepted-by-ase-2016/)   
Yi Li, Chenguang Zhu, Julia Rubin and Marsha Chechik   
In Proceedings of the 31st IEEE/ACM International Conference on Automated Software Engineering (ASE 2016)

[Semantic Slicing of Software Version Histories](https://www.ntu.edu.sg/home/yi_li/posts/paper-accepted-by-ase-2015/)  
Yi Li, Julia Rubin and Marsha Chechik  
In Proceedings of the 30th IEEE/ACM International Conference on
Automated Software Engineering (ASE 2015)

[A Dataset for Dynamic Discovery of Semantic Changes in Version Controlled Software Histories](https://www.ntu.edu.sg/home/yi_li/posts/paper-accepted-by-msr-2017/)   
Chenguang Zhu, Yi Li, Julia Rubin and Marsha Chechik   
In Proceedings of the 14th International Conference on Mining Software Repositories (MSR 2017)

[CSlicerCloud: A Web-Based Semantic History Slicing Framework](https://www.ntu.edu.sg/home/yi_li/posts/paper-accepted-by-icse-2018/)   
Yi Li, Chenguang Zhu, Julia Rubin, and Marsha Chechik   
In Proceedings of 40th International Conference on Software Engineering (ICSE 2018)

[FHistorian: Locating Features in Version Histories](ntu.edu.sg/home/yi_li/posts/paper-accepted-by-splc-2017/)    
Yi Li, Chenguang Zhu, Julia Rubin and Marsha Chechik    
In Proceedings of the 21st International Systems and Software Product Line Conference (SPLC 2017)


CHANGELOG.txt
CSlicer

1.0b (2018-07-17)

- [FIX] Now growing hunk dependency set and compute side-effects of those hunk 
  commits until reaching a fixed-point. This has been a known issue for a long
  time and it has not been manifested in any experiments before. The issue was
  first observed in NET-525 and NET-527: the history slice produced by CSlicer
  misses several commits which would provide the compilation dependencies of
  some hunk commits.

- [FIX] A bug in '-t' (verify cherry-picking results) pollutes pom.xml and 
  interfers with the normal cherry-picking process. This issue was observed in
  running 'cslicer -t' on the history slice of IO-173.

1.0a (2018-07-12)

-----------------

- [NEW] Count and output the number of changed lines at the AST atomic change
  level. This is a very rough over-approximation -- the length of the whole
  method is counted if the method body is modified (even if a single line gets
  changed).

- [FIX] Enable whole class call graph analysis (generating dependencies between
  all classes, methods and fields). A partial call graph analysis was used
  before and this resulted in some missing compilation dependencies in CSV-159,
  CSV-175, and CSV-180. The history slices produced for those examples cannot be
  compiled due to this bug.
  Reported by Chenguang on Jun 11, 2018.

- [FIX] When running the configuration 'cslicer-split-cslicer', CSlicer reuses
  Jacoco files generated for the first stage history. Now we simply reuse the
  functional/compilation set generated at the first stage in the second stage.

=================

Definer

1.0b (2018-08-12)

- [NEW] Implement multi-round Definer. Run Definer loop multiple times.
  Subsequent Definer runs avoid combinations which have already been tried
  before and won't try them again.

1.0a (2018-07-09)

- [FIX] A bug which prevents Definer from getting 1-minimal slice was observed
  in IO-288 and a few others. The issue was that the 'target' directory which
  contains the compiled source code and test code was purged after running
  Definer for a while. The cause of this issue is still unknown (possibly due to
  calling 'reset HEAD --hard' at some point). The solution was to back up the
  compile test code in the test script and recover them when the 'target' folder
  is found empty. 

- [FIX] When manually compiling the test suite, 'mvn test' was used. This
  command would compile and run all the tests once and this time was counted
  towards the total running time of Definer. Now we only run necessary tests in
  the slicing criteria.

=================

Experiment

2018-07-18

- cslicer-split-definer still times out on NET-525 and NET-527. One reason is
  that there is a commmit which introduces bad pom syntax. A lot of the
  partitions are ill-formed due to this reason. We would like to restore pom.xml
  to the correct version before running every test.

2018-07-15

- cslicer-split-definer times out on NET-525 and NET-527. The history slice
  produced after the first stage misses several commits due to a known issue.
  This was then fixed in 1.0b.


resources/scripts/paper/tables.py
#!/usr/bin/python3

import os
import sys
import re
import csv
import time
import argparse
import shutil
import subprocess as sub

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__)) # Dir of this script
_DOWNLOADS_DIR = SCRIPT_DIR + '/../../../_downloads'
OUTPUT_DIR = SCRIPT_DIR + '/../../file-level/output'
CONFIGS_DIR = SCRIPT_DIR + '/../../file-level/definer-configs'
ALL_CONFIGS_NUMBERS_TEX_FILE = SCRIPT_DIR + \
                               '/../../file-level/results/tables/allconfigs-numbers.tex'
EFFECTIVENESS_NUMBERS_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/effectiveness-exp-numbers.tex'

PROJECTS_NUMBERS_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/projects-numbers.tex'
PROJECTS_TABLE_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/projects-table.tex'
SUBJECTS_NUMBERS_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/subjects-numbers.tex'
SUBJECTS_TABLE_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/subjects-table.tex'
SCHEME_NUMBERS_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/scheme-numbers.tex'
SCHEME_TABLE_TEX_FILE = SCRIPT_DIR + \
                    '/../../../../ASEJ2018_SemanticSlicing/tables/scheme-table.tex'

EXAMPLES = ['COMPRESS-327', 'COMPRESS-369', 'COMPRESS-373', 'COMPRESS-374', 'COMPRESS-375',
            'IO-173', 'IO-275', 'IO-288', 'IO-290', 'IO-305',
            'LANG-993', 'LANG-1006',
            'NET-525', 'NET-527',
            'CSV-159', 'CSV-175', 'CSV-180']

PROJECTS = ['compress', 'io', 'lang', 'net', 'csv']

PARTITION_SCHEMES = ['definer-neg', 'definer-nopos', 'definer-low3', 'definer']

DESCRIPTIONS = {'COMPRESS-327': 'Support in-memory processing for ZipFile',
                'COMPRESS-369': 'Allow archiver extensions through a standard JRE ServiceLoader',
                'COMPRESS-373': 'Support writing the \"old\" lzma format',
                'COMPRESS-374': 'Add support for writing lzma streams in 7z',
                'COMPRESS-375': 'Allow the clients of ParallelScatterZipCreator to provide ZipArchiveEntryRequestSupplier',
                'IO-173': 'FileUtils.listFiles() doesn\'t return directories',
                'IO-275': 'Add option to ignore line endings',
                'IO-288': 'Supply a ReversedLinesFileReader',
                'IO-290': 'Add read/readFully methods to IOUtils',
                'IO-305': 'New copy() method in IOUtils that takes additional offset, length and buffersize arguments',
                'LANG-883': 'Add StringUtils.containsAny(CharSequence, CharSequence...) method',
                'LANG-993': 'Add zero copy write method to StrBuilder',
                'LANG-1006': 'Add wrap (with String or char) to StringUtils',
                'LANG-1080': 'Add NoClassNameToStringStyle implementation of ToStringStyle',
                'LANG-1093': 'Add ClassUtils.getAbbreviatedName',
                'NET-525': 'Added missing set methods on NTP class and interface',
                'NET-527': 'Add SimpleNTPServer as example and for testing',
                'CSV-159': 'Add IgnoreCase option for accessing header names',
                'CSV-175': 'Support for ignoring trailing delimiter',
                'CSV-180': 'Add withHeader(Class<? extends Enum>) to CSVFormat'}

def countNumOfJavaFiles(project_dir):
    counter = 0
    for path, subpaths, files in os.walk(project_dir, False):
        for f in files:
            if f.endswith('.java'):
                counter += 1
    return counter

def countLOC(project_dir):
    p = sub.Popen('sloccount ' + project_dir, shell=True, stdout=sub.PIPE, stderr=sub.PIPE)
    p.wait()
    lines = p.stdout.readlines()
    for i in range(len(lines)):
        lines[i] = lines[i].decode("utf-8")[:-1]
        if lines[i].strip().startswith('java:') and lines[i].strip().endswith(')'):
            loc = lines[i].split()[1]
    return int(loc)

def countNumOfCommitsInOneYearAndFourMonths(project_dir):
    cwd = os.getcwd()
    os.chdir(project_dir)
    p = sub.Popen('git --no-pager log --since \"JAN 1 2015\" --until \"JAN 1 2016\" --pretty=format:\"%h %an %ad\"', shell=True, stdout=sub.PIPE, stderr=sub.PIPE)
    p.wait()
    commits = p.stdout.readlines()
    num_of_commits_in_one_year = len(commits)
    num_of_commits_in_four_months = float(num_of_commits_in_one_year) / 4
    os.chdir(cwd)
    return num_of_commits_in_one_year, round(num_of_commits_in_four_months, 2)

def genProjectNumbers(projects=PROJECTS, projects_numbers_tex_file=PROJECTS_NUMBERS_TEX_FILE, \
                      downloads_dir=_DOWNLOADS_DIR):
    lines = ''
    for project in projects:
        project_dir = downloads_dir + '/commons-' + project + '-fake'
        num_of_java_files = countNumOfJavaFiles(project_dir)
        loc = countLOC(project_dir)
        num_of_commits_in_one_year, num_of_commits_in_four_months = \
                            countNumOfCommitsInOneYearAndFourMonths(project_dir)
        lines += '\\DefMacro{' + project + 'NumOfJavaFiles}{' + \
                 '{:,}'.format(num_of_java_files) + '}\n'
        lines += '\\DefMacro{' + project + 'LOC}{' + '{:,}'.format(loc) + '}\n'
        lines += '\\DefMacro{' + project + 'NumOfCommitsOneYear}{' + \
                 '{:,}'.format(num_of_commits_in_one_year) + '}\n'
        lines += '\\DefMacro{' + project + 'NumOfCommitsFourMonths}{' + \
                 '{0:,.2f}'.format(num_of_commits_in_four_months) + '}\n'
    fw = open(projects_numbers_tex_file, 'w')
    fw.write(lines)
    fw.close()

def genProjectTable(projects=PROJECTS, projects_table_tex_file=PROJECTS_TABLE_TEX_FILE):
    table_lines = ''
    table_lines += '\\begin{table}[t]\n'
    table_lines += '\\caption{\\TCaptionProjects}\n'
    table_lines += '\\centering\n'
    #table_lines += '\\resizebox{.4\columnwidth}{!}{\n'
    table_lines += '\\begin{small}\n'
    table_lines += '\\begin{tabular}{lrr}\n'
    table_lines += '\\toprule\n'
    #table_lines += '\\TableHeadProjects & \\TableHeadNumOfFiles & \\TableHeadLOC & \\TableHeadNumOfCommitsOneYear & \\TableHeadNumOfCommitsFourMonths \\\\\n'
    table_lines += '\\TableHeadProjects & \\TableHeadNumOfFiles & \\TableHeadLOC \\\\\n'
    table_lines += '\\midrule\n'
    for project in projects:
        table_lines += '\\' + project
        table_lines += ' & \\UseMacro{' + project + 'NumOfJavaFiles}'
        table_lines += ' & \\UseMacro{' + project + 'LOC}'
        #table_lines += ' & \\UseMacro{' + project + 'NumOfCommitsOneYear}'
        #table_lines += ' & \\UseMacro{' + project + 'NumOfCommitsFourMonths}'
        table_lines += ' \\\\\n'
    table_lines += '\\bottomrule\n'
    table_lines += '\\end{tabular}\n'
    table_lines += '\\end{small}\n'
    table_lines += '\\end{table}\n'
    fw = open(projects_table_tex_file, 'w')
    fw.write(table_lines)
    fw.close()

def extractInfoFromAllConfigNumbersTexFile(example, \
                                    all_configs_numbers_tex_file=ALL_CONFIGS_NUMBERS_TEX_FILE):
    fr = open(all_configs_numbers_tex_file, 'r')
    lines = fr.readlines()
    fr.close()
    for i in range(len(lines)):
        if lines[i].startswith('\\DefMacro{' + example.replace('-', '') + \
                               'definerNumOfCommitsMapBack'):
            hstar = lines[i].split('{')[2].split('}')[0]
        elif lines[i].startswith('\\DefMacro{' + example.replace('-', '') + 'OrigHisLen'):
            h = int(lines[i].split('{')[2].split('}')[0])
    return hstar, h

def extractInfoFromConfigFile(example, configs_dir=CONFIGS_DIR):
    config_file = configs_dir + '/' + example + '.properties'
    fr = open(config_file, 'r')
    lines = fr.readlines()
    fr.close()
    for i in range(len(lines)):
        if lines[i].startswith('endCommit = '):
            end_point = lines[i].strip().split()[-1]
        elif lines[i].startswith('testScope = '):
            num_of_test_methods = lines[i].count('#') + lines[i].count('+')
    return end_point, num_of_test_methods

def genSubjectsNumbers(examples=EXAMPLES, descriptions=DESCRIPTIONS, \
                       subjects_numbers_tex_file=SUBJECTS_NUMBERS_TEX_FILE):
    lines = ''
    for example in examples:
        lines += '\\DefMacro{' + example.replace('-', '') + 'IssueKey}{' + \
                 example + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'Descriptions}{' + \
                 descriptions[example] + '}\n'
        hstar, h = extractInfoFromAllConfigNumbersTexFile(example)
        end_point, num_of_test_methods = extractInfoFromConfigFile(example)
        lines += '\\DefMacro{' + example.replace('-', '') + 'EndPoint}{' + \
                 end_point[:7] + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'AbsHisSize}{' + \
                 '{:,}'.format(h) + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'AbsTestSize}{' + \
                 '{:,}'.format(num_of_test_methods) + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'HstarSize}{' + \
                 hstar + '}\n'
    fw = open(subjects_numbers_tex_file, 'w')
    fw.write(lines)
    fw.close()

def genSubjectsTable(examples=EXAMPLES, subjects_table_tex_file=SUBJECTS_TABLE_TEX_FILE):
    table_lines = ''
    table_lines += '\\begin{table*}[t]\n'
    table_lines += '\\caption{\\TCaptionSubjects}\n'
    # \label{tab.subjects}
    table_lines += '\\centering\n'
    table_lines += '\\resizebox{\\textwidth}{!}{\n'
    table_lines += '\\begin{tabular}{l|llL{8cm}l|rrr}\n'
    table_lines += '\\toprule\n'
    table_lines += '\\TableHeadProjects & \\TableHeadSubjectID & \\TableHeadIssueKey & \\TableHeadDesc & \\TableHeadEndPoint & \\TableHeadAbsHisSize & \\TableHeadAbsTestSize & \\TableHeadHstarSize \\\\\n'
    table_lines += '\midrule\n'
    for example in examples:
        if example == 'LANG-883':
            table_lines += '\midrule\n'
            table_lines += '\\multirow{5}{*}{\\lang}'
        elif example == 'IO-173':
            table_lines += '\midrule\n'
            table_lines += '\\multirow{5}{*}{\\io}'
        elif example == 'COMPRESS-327':
            table_lines += '\\multirow{5}{*}{\\compress}'
        elif example == 'CSV-159':
            table_lines += '\midrule\n'
            table_lines += '\\multirow{3}{*}{\\csv}'
        elif example == 'NET-525':
            table_lines += '\midrule\n'
            table_lines += '\\multirow{2}{*}{\\net}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'ID}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'IssueKey}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'Descriptions}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'EndPoint}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'AbsHisSize}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'AbsTestSize}'
        table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'HstarSize}'
        table_lines += ' \\\\\n'
    table_lines += '\\bottomrule\n'
    table_lines += '\\end{tabular}}\n'
    table_lines += '\\end{table*}\n'
    fw = open(subjects_table_tex_file, 'w')
    fw.write(table_lines)
    fw.close()

def extractInfoFromPartitionNumbers(example, partition_schemes=PARTITION_SCHEMES, \
                                    output_dir=OUTPUT_DIR):
    scheme_tests_dict = {}
    for scheme in partition_schemes:
        fr = open(output_dir + '/' + scheme + '/' + example + '.log', 'r')
        lines = fr.readlines()
        fr.close()
        pattern = re.compile('.* at Iteration .* after .* test.*.$')
        for i in range(len(lines)):
            if re.match(pattern, lines[i].strip()):
                num_of_tests = lines[i].strip().split()[-2]
        scheme_tests_dict[scheme] = num_of_tests
    return scheme_tests_dict

def genPartitionSchemeNumbers(examples=EXAMPLES, \
                              scheme_numbers_tex_file=SCHEME_NUMBERS_TEX_FILE):
    lines = ''
    for example in examples:
        scheme_tests_dict = extractInfoFromPartitionNumbers(example)
        lines += '\\DefMacro{' + example.replace('-', '') + 'NegNumOfRuns}{' + \
                 scheme_tests_dict['definer-neg'] + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'NonPosNumOfRuns}{' + \
                 scheme_tests_dict['definer-nopos'] + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'Low3NumOfRuns}{' + \
                 scheme_tests_dict['definer-low3'] + '}\n'
        lines += '\\DefMacro{' + example.replace('-', '') + 'CombinedNumOfRuns}{' + \
                 scheme_tests_dict['definer'] + '}\n'
    fw = open(scheme_numbers_tex_file, 'w')
    fw.write(lines)
    fw.close()

def findMinSchemes(example, scheme_numbers_tex_file=SCHEME_NUMBERS_TEX_FILE):
    fr = open(scheme_numbers_tex_file, 'r')
    lines = fr.readlines()
    fr.close()
    for i in range(len(lines)):
        if lines[i].startswith('\\DefMacro{' + example.replace('-', '') + 'NegNumOfRuns'):
            neg = int(lines[i].split('{')[2].split('}')[0])
        elif lines[i].startswith('\\DefMacro{' + example.replace('-', '') + 'NonPosNumOfRuns'):
            nonpos = int(lines[i].split('{')[2].split('}')[0])
        elif lines[i].startswith('\\DefMacro{' + example.replace('-', '') + 'Low3NumOfRuns'):
            low3 = int(lines[i].split('{')[2].split('}')[0])
        elif lines[i].startswith('\\DefMacro{' + example.replace('-', '') + 'CombinedNumOfRuns'):
            combined = int(lines[i].split('{')[2].split('}')[0])
    min_val = min([neg, nonpos, low3, combined])
    min_schemes = []
    if neg == min_val:
        min_schemes.append('neg')
    if nonpos == min_val:
        min_schemes.append('nonpos')
    if low3 == min_val:
        min_schemes.append('low3')
    if combined == min_val:
        min_schemes.append('combined')
    return min_schemes

def genPartitionSchemeTable(examples=EXAMPLES, scheme_table_tex_file=SCHEME_TABLE_TEX_FILE):
    table_lines = ''
    table_lines += '\\begin{table}[t]\n'
    table_lines += '\\caption{\\TCaptionPartitionSchemes}\n'
    table_lines += '\\centering\n'
    table_lines += '\\resizebox{.5\\columnwidth}{!}{\n'
    table_lines += '\\begin{tabular}{lrrrr}\n'
    table_lines += '\\toprule\n'
    table_lines += '\\TableHeadSubjectID & \\TableHeadNeg & \\TableHeadNonPos & \\TableHeadLowThree & \\TableHeadCombined \\\\\n'
    table_lines += '\\midrule\n'
    for example in examples:
        if example == 'LANG-993':
            table_lines += '\midrule\n'
        elif example == 'IO-173':
            table_lines += '\midrule\n'
        elif example == 'CSV-159':
            table_lines += '\midrule\n'
        elif example == 'NET-525':
            table_lines += '\midrule\n'
        table_lines += '\\UseMacro{' + example.replace('-', '') + 'ID}'
        min_schemes = findMinSchemes(example)
        if 'neg' in min_schemes:
            table_lines += ' & \\textbf{\\UseMacro{' + example.replace('-', '') + \
                           'NegNumOfRuns}}'
        else:
            table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'NegNumOfRuns}'
        if 'nonpos' in min_schemes:
            table_lines += ' & \\textbf{\\UseMacro{' + example.replace('-', '') + \
                           'NonPosNumOfRuns}}'
        else:
            table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'NonPosNumOfRuns}'
        if 'low3' in min_schemes:
            table_lines += ' & \\textbf{\\UseMacro{' + example.replace('-', '') + \
                           'Low3NumOfRuns}}'
        else:
            table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'Low3NumOfRuns}'
        if 'combined' in min_schemes:
            table_lines += ' & \\textbf{\\UseMacro{' + example.replace('-', '') + \
                       'CombinedNumOfRuns}}'
        else:
            table_lines += ' & \\UseMacro{' + example.replace('-', '') + 'CombinedNumOfRuns}'
        table_lines += ' \\\\\n'
    table_lines += '\\bottomrule\n'
    table_lines += '\\end{tabular}}\n'
    table_lines += '\\end{table}\n'
    fw = open(scheme_table_tex_file, 'w')
    fw.write(table_lines)
    fw.close()

if __name__ == '__main__':
    genProjectNumbers()
    genProjectTable()
    examples = ['COMPRESS-327', 'COMPRESS-369', 'COMPRESS-373', 'COMPRESS-374', 'COMPRESS-375',
                'IO-173', 'IO-275', 'IO-288', 'IO-290', 'IO-305',
                'LANG-883', 'LANG-993', 'LANG-1006', 'LANG-1080', 'LANG-1093',
                'NET-525', 'NET-527',
                'CSV-159', 'CSV-175', 'CSV-180']
    genSubjectsNumbers(examples)
    genSubjectsTable(examples)
    genPartitionSchemeNumbers()
    genPartitionSchemeTable()
